src/axom/klee/Geometry.hpp

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#ifndef AXOM_KLEE_GEOMETRY_HPP
+#define AXOM_KLEE_GEOMETRY_HPP
+
+#include <memory>
+#include <string>
+
+#include "axom/klee/Dimensions.hpp"
+#include "axom/klee/Units.hpp"
+
+namespace axom
+{
+namespace klee
+{
+class GeometryOperator;
+
+/**
+ * Properties of a geometric object which can be transformed by operators
+ */
+struct TransformableGeometryProperties
+{
+  Dimensions dimensions;
+  LengthUnit units;
+};
+
+/**
+ * Compare transformable properties for equality.
+ * \param lhs the left-hand-side operand
+ * \param rhs the right-hand-side operand
+ * \return true if and only if all properties are equal
+ */
+bool operator==(const TransformableGeometryProperties &lhs,
+                const TransformableGeometryProperties &rhs);
+
+/**
+ * Compare transformable properties for inequality.
+ * \param lhs the left-hand-side operand
+ * \param rhs the right-hand-side operand
+ * \return false if and only if all properties are equal
+ */
+inline bool operator!=(const TransformableGeometryProperties &lhs,
+                       const TransformableGeometryProperties &rhs)
+{
+  return !(lhs == rhs);
+}
+
+/**
+ * Represents the geometry specified in a Shape.
+ */
+class Geometry
+{
+public:
+  /**
+   * Create a new Geometry object.
+   *
+   * \param startProperties the transformable properties before any
+   * operators are applied
+   * \param format the format of the file
+   * \param path the path of the file
+   * \param operator_ a possibly null operator to apply to the geometry.
+   */
+  Geometry(const TransformableGeometryProperties &startProperties,
+           std::string format,
+           std::string path,
+           std::shared_ptr<GeometryOperator const> operator_);
+
+  /**
+   * Get the format in which the geometry is specified.
+   *
+   * \return the format of the shape
+   */
+  const std::string &getFormat() const { return m_format; }
+
+  /**
+   * Get the path at which to find the specification of the geometry
+   *
+   * \return the path to the geometry file
+   */
+  const std::string &getPath() const { return m_path; }
+
+  /**
+   * Get a GeometryOperator to apply to this geometry. Can be null.
+   *
+   * \return a potentially null operator to apply to the geometry
+   */
+  std::shared_ptr<GeometryOperator const> const &getGeometryOperator() const
```

Comment:
This function could probably return a `const GeometryOperator*` here, since it seems like there's no change in ownership from this call. (Unless I'm missing something and that's the intent)

Reply:
I'll track down the changes this would lead to. I think it makes sense to not expand ownership here, though I'm not sure it would hurt much if someone wanted to hold on to a copy.

Reply:
Unfortunately, this is already being used in production codes at LLNL that have early copies of this library. I think if we want to revisit this, we should do it as part of a new PR that works with those code teams to figure out how to best resolve this. The usages are minimal and could be easily changed, but we don't want to break them without coordinating with them.